### PR TITLE
Refactor underwater mask/effect

### PR DESCRIPTION
**Pull request by daleeidd**
Refactors underwater to greatly improve code reuse downstream. I have moved some of the files around which should be okay since we haven't done a release yet.

We will have two mask shaders downstream as we want to support camera relative rendering on HDRP (and URP in the future). And we will have one underwater effect shader per RP. HDRP will have an additional one for post-processing until it is removed.

With the mask I brought the camera relative code over as it is a minor change and will keep the shared file the same.

With the underwater effect shared file, modifications for SRP will be required which will be done downstream. There will be as many ApplyUnderwaterEffect functions as necessary (HDRP will certainly have its own) in the file and use Unity provided defines (or our own) to conditionally include them.

Also fixes builds stripping the meniscus shader due to keyword difference.

### CODE REVIEW

**Comment by daleeidd** [comment]
Actually... going to make ApplyUnderwaterEffect agnostic between BIRP and URP non SG first.

File: crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterCurtain.shader [hunk 1]
```diff
@@ -212,7 +212,9 @@ Shader "Crest/Underwater Curtain"
 #if _CAUSTICS_ON
 				if (sceneZ01 != 0.0)
 				{
-					ApplyCaustics(view, lightDir, sceneZ, _Normals, true, sceneColour, cascadeData0, cascadeData1);
+					float3 camForward = mul((float3x3)unity_CameraToWorld, float3(0.0, 0.0, 1.0));
```
Comment (huwb) [inline crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterCurtain.shader:215]: There a way to pull out the forward vector directly which avoids a matrix mul. May be worth it as this code runs over many pixels. Can never remember which but it's either mat._m20_m21_m22 or mat._m02_m12_m22 or mat[2]

File: crest/Assets/Crest/Crest/Shaders/Underwater/Resources/UnderwaterEffect.shader [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+				float wt = ComputeMeniscusWeight(uvScreenSpace, mask, horizonPositionNormal, sceneZ);
+
+#if _DEBUG_VIEW_OCEAN_MASK
+				return DebugRenderOceanMask(isOceanSurface, isUnderwater, mask, sceneColour);
+#endif // _DEBUG_VIEW_OCEAN_MASK
+
+				if (isUnderwater)
+				{
+					const half3 view = normalize(input.viewWS);
+					float3 camForward = mul((float3x3)unity_CameraToWorld, float3(0.0, 0.0, 1.0));
```
Comment (huwb) [inline crest/Assets/Crest/Crest/Shaders/Underwater/Resources/UnderwaterEffect.shader:112]: Another matrix mul (see other comment)

File: crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterMaskShared.hlsl [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		// to be flipped (needs to be investigated). Gaps look bad from above surface, and overlaps look bad
+		// from below surface. We want to close gaps without introducing overlaps. A fixed "epsilon" will
+		// either not solve gaps at large distances or introduce too many overlaps at small distances. Even
+		// with scaling, there are still unsolvable overlaps underwater (especially at large distances).
+		// 100,000 (0.00001) is the maximum position before Unity warns the user of precision issues.
+		worldPos.xz = lerp(tileCenterXZ, worldPos.xz, lerp(1.0, 1.01, max(cameraPositionXZ.x, cameraPositionXZ.y) * 0.00001));
+	}
+
+	// Calculate sample weights. params.z allows shape to be faded out (used on last lod to support pop-less scale transitions)
+	const float wt_smallerLod = (1. - lodAlpha) * cascadeData0._weight;
```
Comment (huwb) [inline crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterMaskShared.hlsl:74]: Would you kindly fix what is likely due to my bad habit - float constants should be 1.0 not 1.

**Review by huwb — APPROVED** [review]
Just nitpicks, looks good to me otherwise

File: crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterCurtain.shader (hunk 1 above)
Comment (daleeidd) [reply crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterCurtain.shader]: Thanks. Changed all of them to _m02_m12_m22 which worked.

File: crest/Assets/Crest/Crest/Shaders/Underwater/Resources/UnderwaterEffect.shader [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+			fixed4 Frag (Varyings input) : SV_Target
+			{
+				// We need this when sampling a screenspace texture.
+				UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input);
+
+				float4 horizonPositionNormal; bool isBelowHorizon;
+				GetHorizonData(input.uv, horizonPositionNormal, isBelowHorizon);
+
+				const float2 uvScreenSpace = UnityStereoTransformScreenSpaceTex(input.uv);
+				half3 sceneColour = UNITY_SAMPLE_SCREENSPACE_TEXTURE(_CrestCameraColorTexture, uvScreenSpace).rgb;
```
[the remark this answers is not in the document]
Comment (daleeidd) [reply crest/Assets/Crest/Crest/Shaders/Underwater/Resources/UnderwaterEffect.shader:95]: Required to be mutable. rawOceanDepth merges into rawDepth. And sceneColour gets modified in ApplyCaustics. The others get initialised as out parameters.

**Comment by daleeidd** [comment]
Thanks. I've fixed all the things.

**Pull request merged**